cluster-autoscaler/core/scaledown/legacy/wrapper.go

```diff
@@ -0,0 +1,116 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package legacy
+
+import (
+	"time"
+
+	"k8s.io/autoscaler/cluster-autoscaler/processors/nodes"
+	"k8s.io/autoscaler/cluster-autoscaler/processors/status"
+	"k8s.io/autoscaler/cluster-autoscaler/simulator"
+	"k8s.io/autoscaler/cluster-autoscaler/utils/errors"
+
+	apiv1 "k8s.io/api/core/v1"
+	policyv1 "k8s.io/api/policy/v1beta1"
+	klog "k8s.io/klog/v2"
+)
+
+// ScaleDownWrapper wraps legacy scaledown logic to satisfy scaledown.Planner &
+// scaledown.Actuator interfaces.
+type ScaleDownWrapper struct {
+	sd        *ScaleDown
+	pdbs      []*policyv1.PodDisruptionBudget
+	processor nodes.ScaleDownNodeProcessor
+}
+
+// NewScaleDownWrapper returns a new ScaleDownWrapper
+func NewScaleDownWrapper(sd *ScaleDown, processor nodes.ScaleDownNodeProcessor) *ScaleDownWrapper {
+	return &ScaleDownWrapper{
+		sd:        sd,
+		processor: processor,
+	}
+}
+
+// UpdateClusterState updates unneeded nodes in the underlying ScaleDown.
+func (p *ScaleDownWrapper) UpdateClusterState(allNodes []*apiv1.Node, deletionsInProgress []string, recentEvictions []*apiv1.Pod, pdbs []*policyv1.PodDisruptionBudget, currentTime time.Time) errors.AutoscalerError {
+	p.sd.CleanUp(currentTime)
+
+	var scaleDownCandidates []*apiv1.Node
+	var podDestinations []*apiv1.Node
+
+	// podDestinations and scaleDownCandidates are initialized based on allNodes variable, which contains only
+	// registered nodes in cluster.
+	// It does not include any upcoming nodes which can be part of clusterSnapshot. As an alternative to using
+	// allNodes here, we could use nodes from clusterSnapshot and explicitly filter out upcoming nodes here but it
+	// is of little (if any) benefit.
+
+	if p.processor == nil {
+		scaleDownCandidates = allNodes
+		podDestinations = allNodes
+	} else {
+		var err errors.AutoscalerError
+		scaleDownCandidates, err = p.processor.GetScaleDownCandidates(p.sd.context, allNodes)
```

Comment:
My understanding is that we want to keep the ScaleDownNodeProcessor? If so, why push calling it into UpdateClusterState and not pass podDestinations and scaleDownCandidates as parameters?
The logic for calling processor is presumably the same in every implementation. Even if we won't need pre-filtering processor in new scale-down, the way to do that would be IMO to just not install the processor if new scale-up is used.